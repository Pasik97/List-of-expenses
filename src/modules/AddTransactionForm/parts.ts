import styled from 'styled-components';
import media from 'utils/media';

export const AddTransactionFormWrapper = styled.div`
   color: #1a1a1a;
   font-size: 18px;
   line-height: 21px;
   margin: 24px 0;
`;

export const Row = styled.div`
   margin-bottom: 8px;

   ${media.fromTablet} {
      display: grid;
      grid-template-columns: 3fr 1fr;
   }
`;

export const InputWrapper = styled.div`
   margin: 0 0 8px 0;

   ${media.fromTablet} {
      display: grid;
      grid-template-columns: 1fr 1fr;
   }
`;

export const InputLabel = styled.p`
   white-space: nowrap;
   margin: 0;

   ${media.onlyMobile} {
      margin-bottom: 6px;
   }

   ${media.fromTablet} {
      flex-grow: 1;
   }
`;

export const Input = styled.input`
   color: #1a1a1a;
   font-size: 18px;
   line-height: 21px;
   border: none;
   outline: none;
   border-bottom: 1px solid #dddddd;

   &:focus {
      border-bottom: 1px solid #1a1a1a;
   }

   ${media.fromTablet} {
      flex-grow: 1;
   }
`;

export const AddButton = styled.button`
   color: #1a1a1a;
   font-size: 18px;
   line-height: 21px;
   background: none;
   border: 2px solid #dddddd;
   padding: 6px 24px;

   &:disabled {
      color: #dddddd;
   }

   &:hover, &active {
      border-color: #1a1a1a;
      cursor: pointer;
   }

   ${media.fromTablet} {
      height: fit-content;
      width: fit-content;
      justify-self: flex-end;
   }
`;

export const ValidateInputWrapper = styled.div`
   display: flex;
   flex-direction: column;

   ${media.fromTablet} {
      margin-left: 16px;
      flex-grow: 1;
   }
`;

export const Validation = styled.p`
   margin: 8px 0 0 0;
   color: red;
   font-size: 13px;
   line-height: 16px;
   min-height: 16px;
`;
