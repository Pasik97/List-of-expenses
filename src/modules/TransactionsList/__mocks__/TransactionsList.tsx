const mock = () => '(TransactionsList /)';

export default mock;
