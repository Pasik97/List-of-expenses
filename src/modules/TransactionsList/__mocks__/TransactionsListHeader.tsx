const mock = () => '(TransactionsListHeader /)';

export default mock;
