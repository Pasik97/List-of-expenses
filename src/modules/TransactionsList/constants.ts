export enum SortType {
   Title = 'Title',
   AmountPln = 'AmountPln',
   AmountEur = 'AmountEur',
}

export enum SortDirection {
   Asc = 'Asc',
   Desc = 'Desc',
}

export interface TransactionListHeaderProps {
   currentSortType: SortType;
   currentSortDirection: SortDirection;
   onSortChange: (sortType: SortType, sortDirection: SortDirection) => void;
};
