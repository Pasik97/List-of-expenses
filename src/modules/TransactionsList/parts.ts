import styled from 'styled-components';
import media from 'utils/media';

export const TransactionsListWrapper = styled.div``;

export const Cell = styled.div`
   color: #1a1a1a;
   font-size: 16px;
   line-height: 21px;
   border-right: 1px solid #1a1a1a;
   padding: 6px;
   display: flex;
   align-items: center;

   &:last-of-type {
      border-right: none;
   }
`;

export const HeaderCell = styled(Cell)`
   font-weight: 700;
   background-color: #cccccc;
   
   &:hover, &:active {
      cursor: pointer;
   }
`;

export const Row = styled.div`
   display: grid;
   grid-template-columns: 1fr 1fr 1fr 1fr;
   border: 1px solid #1a1a1a;
   border-top: none;

   &:first-of-type {
      border-top: 1px solid #1a1a1a;
   }

   &:nth-of-type(2n+3) {
      ${Cell} {
         background-color: #eeeeee;
      }
   }

   ${media.fromTablet} {
      display: grid;
      grid-template-columns: 2fr 1fr 1fr 1fr;
   }

   ${media.fromDesktop} {
      display: grid;
      grid-template-columns: 3fr 1fr 1fr 1fr;
   }
`;

export const Label = styled.p`
   margin: 0;
   overflow-wrap: anywhere;
`;

export const DeleteButton = styled.button`
   color: #4d4d4d;
   font-size: 16px;
   line-height: 21px;
   border: none;
   outline: none;
   background: none;

   &:hover, &:active {
      cursor: pointer;
      color: #1a1a1a;
   }
`;
