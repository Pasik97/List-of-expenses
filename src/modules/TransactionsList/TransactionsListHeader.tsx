import React from 'react';
import * as P from './parts';
import { SortDirection, TransactionListHeaderProps, SortType } from './constants';
import { observer } from 'mobx-react';

const TransactionsListHeader: React.FC<TransactionListHeaderProps> = observer(({
   currentSortType,
   currentSortDirection,
   onSortChange,
}) => {
   const isDescDirection = currentSortDirection === SortDirection.Desc;

   return (
      <P.Row>
         <P.HeaderCell
            onClick={() => onSortChange(SortType.Title, isDescDirection ? SortDirection.Asc : SortDirection.Desc)}
         >
            <P.Label>{currentSortType === SortType.Title && (isDescDirection ? 'ðŸ¡‡ ' : 'ðŸ¡… ')}Title</P.Label>
         </P.HeaderCell>
         <P.HeaderCell
            onClick={() => onSortChange(SortType.AmountPln, isDescDirection ? SortDirection.Asc : SortDirection.Desc)}
         >
            <P.Label>{currentSortType === SortType.AmountPln && (isDescDirection ? 'ðŸ¡‡ ' : 'ðŸ¡… ')}Amount (PLN)</P.Label>
         </P.HeaderCell>
         <P.HeaderCell
            onClick={() => onSortChange(SortType.AmountEur, isDescDirection ? SortDirection.Asc : SortDirection.Desc)}
         >
            <P.Label>{currentSortType === SortType.AmountEur && (isDescDirection ? 'ðŸ¡‡ ' : 'ðŸ¡… ')}Amount (EUR)</P.Label>
         </P.HeaderCell>
         <P.HeaderCell>
            <P.Label>Options</P.Label>
         </P.HeaderCell>
      </P.Row>
   )
});

export default TransactionsListHeader;
