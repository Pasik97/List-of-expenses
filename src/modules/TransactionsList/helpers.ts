import { SortDirection, SortType } from "./constants";
import { Transaction } from "stores/transactions/constants";

export const sortFunction = (
   first: number | string,
   second: number | string,
) => first >= second ? 1 : -1;

export const sortTransactions = (
   transactions: Set<Transaction>,
   sortType: SortType,
   sortDirection: SortDirection,
) => {
   const transactionArray = [...transactions];

   switch (sortType) {
      case SortType.Title:
         transactionArray.sort((transactionA, transactionB) => (
            sortDirection === SortDirection.Asc
               ? sortFunction(transactionA.name, transactionB.name)
               : sortFunction(transactionB.name, transactionA.name)
         ));
         break;
      case SortType.AmountPln:
         transactionArray.sort((transactionA, transactionB) => (
            sortDirection === SortDirection.Asc
               ? sortFunction(transactionA.amountPln, transactionB.amountPln)
               : sortFunction(transactionB.amountPln, transactionA.amountPln)
         ));
         break;
      case SortType.AmountEur:
         transactionArray.sort((transactionA, transactionB) => (
            sortDirection === SortDirection.Asc
               ? sortFunction(transactionA.amountEur, transactionB.amountEur)
               : sortFunction(transactionB.amountEur, transactionA.amountEur)
         ));
         break;
   }

   return transactionArray;
}
