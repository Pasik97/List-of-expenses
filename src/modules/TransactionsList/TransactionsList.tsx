import React, { useContext, useState } from 'react';
import { observer } from 'mobx-react';
import * as P from './parts';
import { SortDirection, SortType } from './constants';
import { sortTransactions } from './helpers';
import TransactionsListHeader from './TransactionsListHeader';
import { StoreProviderContext } from 'modules/StoreProvider/StoreProvider';
import { Transaction } from 'stores/transactions/constants';

const TransactionsList: React.FC = observer(() => {
   const { transactions } = useContext(StoreProviderContext);

   const [sortType, setSortType] = useState<SortType>(SortType.Title);
   const [sortDirection, setSortDirection] = useState<SortDirection>(SortDirection.Desc);

   const onSortChange = (sortType: SortType, sortDirection: SortDirection) => {
      setSortType(sortType);
      setSortDirection(sortDirection);
   }

   const onRemove = (transaction: Transaction) => {
      transactions.removeTransaction(transaction);
   }

   const userTransactions: Transaction[] = sortTransactions(
      transactions.transactionsList,
      sortType,
      sortDirection,
   );

   return !userTransactions.length ? null : (
      <P.TransactionsListWrapper>
         <TransactionsListHeader
            currentSortType={sortType}
            currentSortDirection={sortDirection}
            onSortChange={onSortChange}
         />
         {userTransactions.map((transaction) => (
            <P.Row key={transaction.name + transaction.amountPln + transaction.fixedAmountEur}>
               <P.Cell><P.Label>{transaction.name}</P.Label></P.Cell>
               <P.Cell><P.Label>{transaction.amountPln}</P.Label></P.Cell>
               <P.Cell><P.Label>{transaction.fixedAmountEur}</P.Label></P.Cell>
               <P.Cell><P.DeleteButton onClick={() => onRemove(transaction)}>Delete</P.DeleteButton></P.Cell>
            </P.Row>
         ))}
      </P.TransactionsListWrapper>
   );
})

export default TransactionsList;
