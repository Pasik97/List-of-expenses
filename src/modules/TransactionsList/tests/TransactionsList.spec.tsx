import { act, create } from "react-test-renderer";
import * as M from './mockData';
import TransactionsList from "../TransactionsList";
import TransactionsListHeader from "../TransactionsListHeader";
import { SortDirection, SortType } from "../constants";
import { DeleteButton } from "../parts";
import StoreProvider from "modules/StoreProvider/StoreProvider";

jest.mock("../TransactionsListHeader");

describe('<TransactionsList />', () => {
   it('should render TransactionsListHeader and rows with transactions with title, amount pln and eur and button to delete', () => {
      const component = create(
         <StoreProvider store={M.mockStoreWithMultipleTransactions}>
            <TransactionsList />
         </StoreProvider>,
      );

      expect(component.toJSON()).toMatchSnapshot();
   });

   it('should render null if user have no transactions', () => {
      const component = create(
         <StoreProvider store={M.mockStoreWithEmptyTransactionsList}>
            <TransactionsList />
         </StoreProvider>,
      );

      expect(component.toJSON()).toBeNull();
   });

   describe('onSortChange', () => {
      it('should set sortType and sortDirection based on received arguments', () => {
         const component = create(
            <StoreProvider store={M.mockStoreWithMultipleTransactions}>
               <TransactionsList />
            </StoreProvider>,
         );
         let header = component.root.findByType(TransactionsListHeader);

         expect(header.props.currentSortType).toEqual(SortType.Title);
         expect(header.props.currentSortDirection).toEqual(SortDirection.Desc);

         act(() => {
            header.props.onSortChange(SortType.AmountPln, SortDirection.Asc)
         });
         
         header = component.root.findByType(TransactionsListHeader);

         expect(header.props.currentSortType).toEqual(SortType.AmountPln);
         expect(header.props.currentSortDirection).toEqual(SortDirection.Asc);
      });
   });

   describe('onRemove', () => {
      it('should call removeTransaction Transactions class method', () => {
         const component = create(
            <StoreProvider store={M.mockStoreWithMultipleTransactions}>
               <TransactionsList />
            </StoreProvider>,
         );
         const deleteButton = component.root.findAllByType(DeleteButton)[0];

         deleteButton.props.onClick();

         expect(M.mockStoreWithMultipleTransactions.transactions.removeTransaction).toHaveBeenCalled();
      });
   });
});
