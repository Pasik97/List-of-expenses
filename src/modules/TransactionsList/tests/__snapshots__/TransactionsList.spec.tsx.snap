// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TransactionsList /> should render TransactionsListHeader and rows with transactions with title, amount pln and eur and button to delete 1`] = `
<div
  className="sc-bdvvtL"
>
  (TransactionsListHeader /)
  <div
    className="sc-hKwDye jGGgdQ"
  >
    <div
      className="sc-gsDKAQ yMCUW"
    >
      <p
        className="sc-eCImPb hZlHbZ"
      >
        Mockkkkk transaction
      </p>
    </div>
    <div
      className="sc-gsDKAQ yMCUW"
    >
      <p
        className="sc-eCImPb hZlHbZ"
      >
        1
      </p>
    </div>
    <div
      className="sc-gsDKAQ yMCUW"
    >
      <p
        className="sc-eCImPb hZlHbZ"
      >
        3
      </p>
    </div>
    <div
      className="sc-gsDKAQ yMCUW"
    >
      <button
        className="sc-jRQBWg hYKtVa"
        onClick={[Function]}
      >
        Delete
      </button>
    </div>
  </div>
  <div
    className="sc-hKwDye jGGgdQ"
  >
    <div
      className="sc-gsDKAQ yMCUW"
    >
      <p
        className="sc-eCImPb hZlHbZ"
      >
        Mockkkkk tra nsa ction 123
      </p>
    </div>
    <div
      className="sc-gsDKAQ yMCUW"
    >
      <p
        className="sc-eCImPb hZlHbZ"
      >
        123
      </p>
    </div>
    <div
      className="sc-gsDKAQ yMCUW"
    >
      <p
        className="sc-eCImPb hZlHbZ"
      >
        33
      </p>
    </div>
    <div
      className="sc-gsDKAQ yMCUW"
    >
      <button
        className="sc-jRQBWg hYKtVa"
        onClick={[Function]}
      >
        Delete
      </button>
    </div>
  </div>
  <div
    className="sc-hKwDye jGGgdQ"
  >
    <div
      className="sc-gsDKAQ yMCUW"
    >
      <p
        className="sc-eCImPb hZlHbZ"
      >
        Mock transaction
      </p>
    </div>
    <div
      className="sc-gsDKAQ yMCUW"
    >
      <p
        className="sc-eCImPb hZlHbZ"
      >
        12
      </p>
    </div>
    <div
      className="sc-gsDKAQ yMCUW"
    >
      <p
        className="sc-eCImPb hZlHbZ"
      >
        12
      </p>
    </div>
    <div
      className="sc-gsDKAQ yMCUW"
    >
      <button
        className="sc-jRQBWg hYKtVa"
        onClick={[Function]}
      >
        Delete
      </button>
    </div>
  </div>
</div>
`;
